### ISSUE TITLE
Node crashing for Scylla open source version 4.0.0-0.20200513.faf300382ac during repair

### ISSUE THREAD

**Issue by antonio-tringali** [scylladb/scylla#6993]
In a Scylla open source version 4.0.0-0.20200513.faf300382ac a node has crashed with the following logs:

```
ERROR 2020-08-07 03:41:21,188 [shard 20] sstable - writer failed to close file: std::bad_alloc (std::bad_alloc)
scylla: ./seastar/src/core/file.cc:492: virtual seastar::append_challenged_posix_file_impl::~append_challenged_posix_file_impl(): Assertion `_closing_state == state::closed' failed.
INFO  2020-08-07 03:41:21,194 [shard 9] repair - Repair 2878 out of 3054 ranges, id=1, shard=9, keyspace=rdrive_pacificteck, table={versions, tfa_used_recovery_codes, service_config, replication_status, policies, oplog, login, lock_status, live_billing, keys_by_user_id, gocqlx_migrate, emails, config, policies_versions, bucket_tags, bucket_size, bu
cket_logging, billing_space_daily, insights_users, billing, billing_space_monthly, lock_status_v2, users_by_id, serials, whitelist, keys_by_access, billing_stats_daily, sso_saml, customer_counters, bucket_by_name, noncurrent_versions, monitoring, folders, resellers, users_by_name, object_tags, root, billing_stats_monthly, reseller_bucket, multipart_uploads, failed_login, tokens, lifecycle_rules, streaming_object}, range=(7781994160530664406, 7782629590882254353]
Aborting on shard 20.
Backtrace:
0x0000000002edbf92
0x0000000002e818cb
0x0000000002e81bc5
0x0000000002e81c60
0x00007f844d7f7b1f
/opt/scylladb/libreloc/libc.so.6+0x000000000003c624
/opt/scylladb/libreloc/libc.so.6+0x00000000000258d8
/opt/scylladb/libreloc/libc.so.6+0x00000000000257a8
/opt/scylladb/libreloc/libc.so.6+0x0000000000034a65
0x0000000002e0d2d1
0x0000000002e0d31c
0x0000000000e39bf8
0x0000000002e25951
0x00000000012af15b
0x00000000011d5644
0x0000000001295fa4
0x0000000001296428
0x0000000001202784
0x00000000011cc166
0x0000000001192476
0x000000000313e94c

```

Sometimes some shards do not complete repair for other nodes in the cluster.

In general, for some tables some SELECTs are sluggish. If you ask for more than a row or a couple of rows queries tend to timeout.

Sometimes we see rows like:

```
WARN  2020-08-07 07:19:55,824 [shard 36] storage_proxy - Failed to apply mutation from 10.2.12.27#36: std::bad_alloc (std::bad_alloc)
WARN  2020-08-07 07:19:55,829 [shard 36] storage_proxy - Failed to apply mutation from 10.2.12.27#36: std::bad_alloc (std::bad_alloc)
ERROR 2020-08-07 07:19:55,832 [shard 36] storage_proxy - exception during mutation write to 10.7.2.65: std::bad_alloc (std::bad_alloc)

```

These are the logs in the nodes when cqlsh informs that some node did not answer for a query.

The questions are simple and related:

1. Are these known problems for release 4.0.0?

2. Would an upgrade to a later version solve these problems?

3. Which version, if so is the case, would be advised for upgrade?

In a related post, https://github.com/scylladb/scylla/issues/6992, we had problems upgrading!

*Installation details*

Scylla version 4.0.0-0.20200513.faf300382ac with build-id 06b6173cfa08e877d64922e40318826f275

Cluster size: 12 nodes, CentOS 7.8 Docker images

**Comment by asias** [comment]
Do you have MV Tables or Secondary Index tables?

**Comment by asias** [comment]
We need to backport

```
commit f488eaebafe4fdb4355c8ef36770f5807934e3fa
Merge: fe127a215 929cdd3a1
Author: Nadav Har'El <nyh@scylladb.com>
Date:   Sun Jul 26 14:11:41 2020 +0300

    merge: db/view: view_update_generator: make staging reader evictable


commit 5ebe2c28d145cc55d8f8fc86837cef5f5936aa7c
Author: Botond Dénes <bdenes@scylladb.com>
Date:   Mon Jul 6 16:51:08 2020 +0300

    db/view: view_update_generator: re-balance wait/signal on the register semaphore
    
    
    
commit a9c7a1a86c426a235f20d8a742cfdeba56930672
Merge: c2939c67b fbbc86e18
Author: Avi Kivity <avi@scylladb.com>
Date:   Wed Jun 24 12:54:58 2020 +0300

    Merge "repair: row_level: prevent deadlocks when repairing homogenous nodes" from Botond
    
    "
    Row level repair, when using a local reader, is prone to deadlocking on
    the streaming reader concurrency semaphore. This has been observed to
    happen with at least two participating nodes, running more concurrent
    repairs than the maximum allowed amount of reads by the concurrency
    semaphore. In this situation, it is possible that two repair instances,
    competing for the last available permits on both nodes, get a permit on
    one of the nodes and get queued on the other one respectively. As
    neither will let go of the permit it already acquired, nor give up
    waiting on the failed-to-acquired permit, a deadlock happens.
```

for the "repair does not complete" issue.

@anton-dessiatov those patches are related but are not present in 4.0 branch. It will take some time to backport to 4.1 branch. I recommend to upgrade to 4.1 or even 4.2 branch.  

In the meanwhile, I will take a look at #/6992.

**Comment by bhalevy** [comment]
In addition, the following is an instance of https://github.com/scylladb/scylla/issues/6448
```
scylla: ./seastar/src/core/file.cc:492: virtual seastar::append_challenged_posix_file_impl::~append_challenged_posix_file_impl(): Assertion `_closing_state == state::closed' failed.
```

**Comment by slivne** [comment]
4.0.8 include 2708b0d6648739d000f624db2ea05ab86b09286b

@asias do we need to backport to 4.0 the other two patches they are not part of 4.0.8

**Comment by slivne** [comment]
@avikivity / @bhalevy FYI

**Comment by slivne** [comment]
@asias / @bhalevy / @avikivity ping

**Comment by avikivity** [comment]
IIRC @denesb said we don't need the other two for 4.0.

**Comment by slivne** [comment]
@denesb ?

**Comment by denesb** [comment]
Yes, we don't need the other two for 4.0.

**Comment by avikivity** [comment]
Closing as fixed in scylla 4.0.8.
